Fix crash when leader is pushed to next line with reduced width

bertfrees opened the pull request:
@kalaspuffar @PaulRambags A DAISY Pipeline user found a bug in Dotify and I think (hope) I have fixed it. You may already review if you want but I made this a draft PR because I still need to add a test.

bertfrees committed:
Don't silently ignore a PaginatorException

in VolumeProvider.updateVolumeContents(). This will result in a
NullPointerException in VolumeImpl.

bertfrees committed:
Fix failure when leader is pushed to next line with reduced width

PaulRambags commented on src/org/daisy/dotify/formatter/impl/VolumeProvider.java:
```diff
@@ -295,7 +295,7 @@ private SectionBuilder updateVolumeContents(int volumeNumber, ArrayList<AnchorDa
             return sb;
         } catch (PaginatorException e) {
-            return null;
+            throw new RuntimeException(e);
```
Why not simply remove the `try` .. `catch` ? This seems a bit tricky to me.

bertfrees replied:
I guess we could expose PaginatorException in some more places. But what is the point? Then it still needs to be caught further down and converted to some another kind of exception because we don't expose PaginatorException in the [formatter API](https://mtmse.github.io/dotify.api/latest/javadoc/index.html). Check the code, there are several other places where a PaginatorException is converted to a RuntimeException.

PaulRambags commented on src/org/daisy/dotify/formatter/impl/row/SegmentProcessor.java:
```diff
@@ -1102,12 +1097,13 @@ private String getLeftIndentWithLabel(String listLabel, int indent) {
         ) {
-            RowImpl ret = null;
+            if (!hasCurrentRow()) {
+                throw new RuntimeException("Error in code.");
```
I'd prefer a better explanation.

bertfrees replied:
This is basically just to assert that this method is never called when `!hasCurrentRow()`. If this assertion fails it means there is a bug in the code. I'll add a comment.

bertfrees replied:
Added some comments.

PaulRambags commented on src/org/daisy/dotify/formatter/impl/row/SegmentProcessor.java:
```diff
@@ -1127,17 +1123,7 @@ private String getLeftIndentWithLabel(String listLabel, int indent) {
-                    );
-                    //update offset
-                    offset = leaderPos - row.getPreTabPosition(currentRow);
+                    return Optional.ofNullable(flushCurrentRow());
```
Does this also work when the current row is the last row on a page? (and then there should be a flush on a new page)

bertfrees replied:
Not sure what you mean. SegmentProcessor has no notion of pages.

PaulRambags reviewed (changes requested):
This is a simplification of the original code, which is good!

bertfrees committed:
Add comments to the "coding error" exceptions in SegmentProcessor

after Paul's suggestion, see https://github.com/mtmse/dotify.formatter.impl/pull/37#discussion_r538643583.

bertfrees committed:
Make sure that a leader is never separated from the content following it

This change might not actually be needed, or maybe it is only needed
in rare conditions. But in any case it makes the code clearer.

bertfrees committed:
Prevent right-text-indent from stripping spaces before a leader

bertfrees committed:
Add test for right-text-indent with bigger indentation of last row

bertfrees committed:
Support right-text-indent also when trailing content is not right aligned

bertfrees commented:
@kalaspuffar @PaulRambags I added more commits. Here is a summary of what this PR contains now:

- Fixed a failure when a leader flows into a footer line with reduced width and a soft wrap happened right before it (so the previous row has the full width available). This was my initial commit but I found several more bugs after that, see the other items below. A test has also been added now.
- Support right-text-indent also when the content after the leader is not right aligned. I don't know why I didn't support that before, it is a use case that makes perfect sense.
- Prevent right-text-indent from stripping spaces before a leader.

By the way I found yet another bug but haven't had the time to fix it yet. (It's not trivial.) See https://github.com/mtmse/dotify.formatter.impl/issues/41.

PaulRambags commented:
@bertfrees Checkstyle is unhappy with the position of a '}'. Please fix that.

PaulRambags commented on integrationtest/org/daisy/dotify/formatter/test/resource-files/leader-right-text-indent-expected.pef:
```diff
@@ -6,16 +6,36 @@
 			<dc:date>2020-06-04</dc:date>
 			<dc:title>Right text indent</dc:title>
-			<dc:description>Tests a right-aligned leader on the last line of a block which is indented on the right side for all the other lines.</dc:description>
+			<dc:description>Tests a leader on the last line of a block that is indented on the right side for all the other lines. Tests both right alignment and left aligment. Tests both with a leader that is aligned to the right edge of the page and with a leader that is aligned to a position that is further than 'right-text-indent' cells away from the right edge of the page.</dc:description>
```
This line is very long.

bertfrees replied:
It is indeed. However it's the case for all tests. Maybe for a reason, I'm not sure. Maybe it is needed to generate the tests-overview.xml file. I suggest we fix this for all tests at once in a separate patch.

PaulRambags commented on src/org/daisy/dotify/formatter/impl/row/SegmentProcessor.java:
```diff
@@ -1115,51 +1114,38 @@ private String getLeftIndentWithLabel(String listLabel, int indent) {
+            if (!hasCurrentRow()) {
+                // continueRow() should only be called after having checked that there is a current
+                // row, or after having created a new one.
+                throw new RuntimeException("Error in code.");
```
Personally I don't like such a vague error message.
`throw new RuntimeException("Method continueRow was called but there is no current row.");`

bertfrees replied:
OK I can see where you're coming from, but it's just meant as an assertion. The message is not meant as a user-level error message. I guess I could change it into an `assert`, but the reason I usually not do that is because assertions are disabled by default. Do you think we should use real assertions?

PaulRambags replied:
I added "Personally" because it is partly a matter of taste. "Error in code" is just so meaningless. I've encountered this message several times - and there were cases where I think I didn't do anything wrong. A more descriptive explanation helps when you need to determine what went wrong. Of course you can still check the stack trace.

bertfrees replied:
Yes, checking the stack trace and the source is the idea.

If you read up on using assertions, they also recommend not providing a message unless "the program has some additional information that might help diagnose the failure" (see https://docs.oracle.com/javase/7/docs/technotes/guides/language/assert.html).

If one of these error messages ever gets exposed to users because of a bug, I think something like "Coding error" makes it clear that they have to contact the developers. Any additional information can only be confusing for the user because it is not meant for them.

> I've encountered this message several times - and there were cases where I think I didn't do anything wrong.

What do you mean?

PaulRambags commented on src/org/daisy/dotify/formatter/impl/row/SegmentProcessor.java:
```diff
@@ -1191,12 +1199,21 @@ private String getLeftIndentWithLabel(String listLabel, int indent) {
             // don't know if soft hyphens need to be replaced, but we'll keep it for now
             next = softHyphenPattern.matcher(next).replaceAll("");
-            if ("".equals(next) && "".equals(tabSpace)) {
+            // If there is a leader, insert it if the content that follows it fits on the line or if
```
Please remove "it" in "that follows it fits"

PaulRambags commented:
@bertfrees Please fix the checkstyle error

bertfrees commented:
@PaulRambags I made checkstyle happy. (I fixed up the offending commit and force-pushed.)

@kalaspuffar You asked me to remind you to not squash the commits this time. The different commits address different issues, so I think it makes sense to not squash. I can rebase onto master if needed.

kalaspuffar commented:
Hi

Running regression tests.

Best regards
Daniel
